### ISSUE TITLE
vscode hanging with some vue files

### ISSUE THREAD

**Issue by gurgeous** [tailwindlabs/prettier-plugin-tailwindcss#144]
**What version of `prettier-plugin-tailwindcss` are you using?**

0.2.5

**What version of Tailwind CSS are you using?**

3.2.7

**What version of Node.js are you using?**

18.12.1

**What package manager are you using?**

yarn

**What operating system are you using?**

macOS

**Describe your issue**

We added `prettier-plugin-tailwindcss` a couple of days ago. It worked initially, but I noticed that vscode+prettier was hanging trying to save certain `.vue` files. The hangs were  obvious because files would fail to save and vscode would complain that prettier was still running. This occurred even after restarting vscode, etc. I removed the package and vscode was able to save those files again.

Sorry about the vague bug report. We are not using any other prettier plugins. We love the idea and really want it for our project! Thanks for all your hard work.

**Comment by thecrypticace** [comment]
Does prettier hang when you run `npx prettier -w .` in your project?

**Comment by gurgeous** [comment]
I cannot get it to happen with prettier on the CLI. It's intermittent in vscode, but definitely present and unrecoverable. Here's the shortest file I could make that triggers it occasionally:

Something.vue:
```vue
<template>
  
  <div v-if="error" class="bg-error text-xs mt-1 bg-error-content py-1 px-2 inline-block rounded">
    {{ error }}
  </div>
</template>

<script setup lang="ts">
import { ref } from "vue";
const error = ref<string>();
</script>
```

```sh
$ code --list-extensions | rg "vue|prett"
esbenp.prettier-vscode
Vue.volar
Vue.vscode-typescript-vue-plugin
```

settings.json:
```
  "editor.formatOnSave": true,
  "[vue]": {
    "editor.codeActionsOnSave": { "source.organizeImports": true },
    "editor.defaultFormatter": "esbenp.prettier-vscode"
  },
```

.prettierrc:
```
{
  "printWidth": 100
}
```

**Comment by thecrypticace** [comment]
Are you by chance using the Tailwind CSS intellisense extension as well?

**Comment by gurgeous** [comment]
I am not using that. Anything else I should grep for? Thanks!!

**Comment by thecrypticace** [comment]
Nope, I don't think so. I'll see what I can figure out.

**Comment by thecrypticace** [comment]
@gurgeous So far I can't get the hang to happen. Any chance you could provide a project that you're able to reproduce it on? Also the full list of install extensions would be super helpful.

**Comment by gurgeous** [comment]
I can't share the project unfortunately. Maybe I can come up with a better repro. I'm also happy to pop open dev tools or mess around with vscode if you have any pointers. I've written some extensions too...

Here's the full list of installed extensions, apologies:

aki77.haml-lint alexcvzz.vscode-sqlite andersliu.insert-line-number astro-build.astro-vscode bierner.markdown-preview-github-styles BriteSnow.vscode-toggle-quotes bungcip.better-toml coolbear.systemd-unit-file cpylua.language-postcss csstools.postcss dbaeumer.vscode-eslint donjayamanne.githistory doublefint.pgsql earshinov.permute-lines EditorConfig.EditorConfig enkia.tokyo-night esbenp.prettier-vscode flesler.url-encode formulahendry.code-runner GitHub.copilot GitHub.vscode-pull-request-github golang.go Grumpydev.pico8vscodeeditor gurgeous.ruby-open-gem hogashi.crontab-syntax-highlight idleberg.applescript JuanBlanco.solidity kamikillerto.vscode-colorize karunamurti.haml Koihik.vscode-lua-format LoranKloeze.ruby-rubocop-revived mechatroner.rainbow-csv misogi.ruby-rubocop mrmlnc.vscode-duplicate ms-azuretools.vscode-docker ms-python.isort ms-python.python ms-python.vscode-pylance ms-vscode-remote.remote-ssh ms-vscode-remote.remote-ssh-edit ms-vscode.remote-explorer oderwat.indent-rainbow Orta.vscode-jest Prisma.prisma rebornix.ruby sgryjp.vscode-stable-sort Shopify.ruby-lsp skellock.just sldobri.gruvbox-5-stars sryze.uridecode stkb.rewrap stylelint.vscode-stylelint svelte.svelte-vscode syler.sass-indented tiehuis.zig Vue.volar Vue.vscode-typescript-vue-plugin william-voyek.vscode-nginx wingrunr21.vscode-ruby zkirkland.vscode-firstupper

**Comment by kotaksempit** [comment]
Me too having the same issue on vscode. Had to removed this plugin to make saving vue file works again.

**Comment by gurgeous** [comment]
We've adopted a workaround for anyone interested:

1. Set `pluginSearchDirs: false` in `.prettierrc`. This fixes vscode.
2. Run `prettier --plugin node_modules/prettier-plugin-tailwindcss` from the CLI when you want to sort things.

We use `just` as our command runner and added a `just sort` to run the above. You could also use a pre-commit hook, etc.

Again, thanks for the great plugin! We love it (clearly) and  I totally get that vscode/prettier are tough to debug. No worries and thanks for all your hard work.

**Comment by tjkohli** [comment]
For what it's worth, we're having the same issue with Next.js + TypeScript + TailwindCSS files. After spending over 4 hours trying to isolate the root cause, it was this plugin. Removing it solved the issue!

**Comment by ddahan** [comment]
After days on a similar issue where one my CPU was stuck at 100% usage, I can confirm this comes from this plugin!

I'm using Nuxt 3.41 and I'm formatting Vue files.
On a 100-line Vue file:
- Saving a file with Prettier only: 60ms and no CPU issue
- Saving a file with Prettier + prettier-plugin-tailwind: 400ms, and minutes later, CPU usage raises, then my VS Code is completely stuck, and I have to reboot it.

Uninstalled :(
Too bad, it was useful to me.

**Comment by tkat** [comment]
In vue-cli based projects,  downgrading Tailwind to 3.2.x resolves the issue.

**Comment by AjayDevInfy** [comment]
Uninstall solve this :(

**Comment by mingtheanlay** [comment]
Any update on this? it seems pretty bad right now

**Comment by aaronmw** [comment]
You can run "Restart Extension Host" in VSCode to avoid rebooting the whole app and project — it seems to halt the runaway process and gets things working again. Sometimes I can just re-save the file afterwards without crash, but then a few more saves of the same file and it's crashing again. 

Maybe that's a clue to what's happening with the plugin? I added `pluginSearchDirs: false` to `prettier.config.js` in my NextJS + Typescript project (as per @gurgeous' comment above)  and I haven't had the issue so far. Though it's only been minutes, the one file that was crashing consistently seems not to be anymore ¯\_(ツ)_/¯

**Comment by thecrypticace** [comment]
👋 We're working towards a solution for this.

I've merged some (sometimes significant) performance and memory improvements in #153 that could potentially fix this problem. The fix is available to test via our insiders build: `npm install prettier-plugin-tailwindcss@insiders`

Could some of you give it a test and report back? It would be super helpful! You'll need to close/reopen VS Code after installing it so the prettier extension doesn't have the old version in memory.

**Comment by ddahan** [comment]
@thecrypticace installing the plugin broke my Prettier, I have no formatting anymore when saving a file. Prettier logs when saving a file:

```
["INFO" - 3:35:55 PM] Formatting file:///workspaces/ppw/neuxt/pages/toolbox.vue
["INFO" - 3:35:55 PM] Using ignore file (if present) at /workspaces/ppw/.prettierignore
```

Notes:
- I use Prettier for VS code extension
- I have no `.prettierignore` file
- I have no specific settings that could explain this.
- I uninstalled the plugin and get a normal behaviour back, here are logs for a single save action:

```
["INFO" - 3:41:16 PM] Formatting file:///workspaces/ppw/neuxt/pages/toolbox.vue
["INFO" - 3:41:16 PM] Using ignore file (if present) at /workspaces/ppw/.prettierignore
["INFO" - 3:41:16 PM] File Info:
{
  "ignored": false,
  "inferredParser": "vue"
}
["INFO" - 3:41:16 PM] No local configuration (i.e. .prettierrc or .editorconfig) detected, falling back to VS Code configuration
["INFO" - 3:41:16 PM] Prettier Options:
{
  "arrowParens": "always",
  "bracketSpacing": true,
  "endOfLine": "lf",
  "htmlWhitespaceSensitivity": "css",
  "insertPragma": false,
  "singleAttributePerLine": true,
  "bracketSameLine": false,
  "jsxBracketSameLine": false,
  "jsxSingleQuote": false,
  "printWidth": 90,
  "proseWrap": "preserve",
  "quoteProps": "as-needed",
  "requirePragma": false,
  "semi": true,
  "singleQuote": true,
  "tabWidth": 2,
  "trailingComma": "es5",
  "useTabs": false,
  "vueIndentScriptAndStyle": false,
  "filepath": "/workspaces/ppw/neuxt/pages/toolbox.vue",
  "parser": "vue"
}
["INFO" - 3:41:16 PM] Formatting completed in 69ms.
```

**Comment by thecrypticace** [comment]
@ddahan There was a small hiccup that I just pushed a fix for a few minutes ago. Can you re-install the insiders build and give it one more test please?

The insiders version you'll need to test is `0.0.0-insiders.78bd35b` (you can check your lock file to see if its the right one)

**Comment by ddahan** [comment]
@thecrypticace it is definitely better than before! 
Before, with the plugin activated: 400ms instead of 50-100ms to save a Vue file of 100 lines (and 100% CPU usage).
Now it seems to be almost the same.

However, it feels like if I wait for a few seconds to save the file, it takes longer (around 250ms). Then it's quick again. Is there anything that could explain this difference?

**Comment by thecrypticace** [comment]
Yep so we removed object hashing to speed up config loading as for some config files it becomes very expensive but added an expiration timer so we can still detect changes to the config file (otherwise you'd have to reload VS Code every time the config changed). I'll see if there are other options for detecting config file changes to reduce the need for shorter expiration times.

**Comment by ddahan** [comment]
Ok thanks! Not sure to understand deeply what's happening behind the scene, but the DX is a little weird because of that, because for the same file you have an almost instantaneous formatting, and then a laggy one. Anything that could improve this would be welcome imo. Thanks!

**Comment by Doesntmeananything** [comment]
Still getting the same issue even with the insiders build `0.0.0-insiders.78bd35b`. Disabling formatting on save with Prettier completely eliminates the issue, so it seems that the underlying issue is still there.

Not sure how to troubleshoot this further, but would be happy to provide any information that could be helpful here.

**Comment by valgeirb** [comment]
> Still getting the same issue even with the insiders build `0.0.0-insiders.78bd35b`. Disabling formatting on save with Prettier completely eliminates the problem, so it seems that the underlying cause is still there.
> 
> Not sure how to troubleshoot this further, but would be happy to provide any information that could be helpful here.

Same thing for me with `0.0.0-insiders.78bd35b`.

**Comment by thecrypticace** [comment]
@Doesntmeananything @valgeirb Could either one or both of you provide reproduction projects , details on VSCode plugins and VSCode version, and info on how you reproduce it? A video / screen recording might be helpful too so we can see what you're doing to repro it.

Additionally, specs on your computer could be useful (like CPU, memory, etc…)

**Comment by krisz094** [comment]
I'm having the same issue and I'm rather curious what the common denominator is between people who have this.

I think this is related to a combination of installed NPM packages, because I've tried different VS Code versions, different Prettier extension versions, used different computers (one was a brand new Win11 installation), etc. 
But there is just one project that is giving me issues while others work flawlessly.

Are any of you using DaisyUI for example?

Meanwhile I'm trying to create a minimal repro example.

**Comment by thecrypticace** [comment]
@krisz094 Even if the repro isn't minimal it might still be useful as long as the issue is reproducible. Is that repo public by chance?

**Comment by kmcgurty** [comment]
FWIW, I'm using `Sveltekit 1.15.8` + `Tailwindcss 3.2.4` + `Typescript 5.0.4` + this plugin. The insiders version `0.0.0-insiders.d3f787d` made a significant improvement. When VSCode upgraded to [Typescript 5](https://code.visualstudio.com/updates/v1_77#_typescript-50) last month, that's when we noticed the performance hit when autosaving. The workaround at the time was to downgrade VSCode to a patch before 1.77.

**Comment by or2e** [comment]
We are reproducing
https://github.com/verdie-g/crpg/tree/master/src/WebUI

**Comment by jd1378** [comment]
can't we watch config file and update instead of expiring it after x seconds ?

**Comment by thecrypticace** [comment]
@jd1378 
> can't we watch config file and update instead of expiring it after x seconds ?

We chose this router because, as far as I know, Prettier does not offer an API to do this. This means that it would need to be done via Node APIs — possibly using chokidar or some other file watcher. This also means that if we did that then we'd likely need to detect that we're running in VSCode because there's overhead to doing that and you wouldn't want to do it when running prettier from the command line.

**Comment by jd1378** [comment]
that makes sense. but can't we detect and only do it when running in vscode ? so we don't get the overhead in cli, but get the benefits in vscode ? also it should be only one file I guess, so the overhead should be small. I mean if it means we don't reload the config for the rest of the life time of the extension, that would be a lot less work

------------
I'm on a nuxt 3 project, it it not big yet, but still has ~ 50+ components already

here's without `prettier-plugin-tailwindcss`:
```
Rule                              | Time (ms) | Relative
:---------------------------------|----------:|--------:
prettier/prettier                 |  1735.189 |    50.0%
import/namespace                  |  1262.053 |    36.4%
@typescript-eslint/no-unused-vars |    70.073 |     2.0%
vue/attributes-order              |    63.656 |     1.8%
n/no-deprecated-api               |    25.548 |     0.7%
no-redeclare                      |    14.051 |     0.4%
import/order                      |    13.544 |     0.4%
unicorn/escape-case               |    12.815 |     0.4%
vue/component-tags-order          |     9.894 |     0.3%
spaced-comment                    |     9.673 |     0.3%
```

**Comment by jd1378** [comment]
also as a thought: after squeezing as much as performance we can, can't we then create some web workers and offload the processing to web workers to process expressions in parallel ? assuming each expression must be done on it's own, wouldn't that help ?

**Comment by thecrypticace** [comment]
@jd1378 While it would be amazing if that were possible — processing things in parallel in web workers would unfortunately require prettier integration. JS/TS expressions in non-JS/JSX/TS/TSX files are usually handled as embedded documents which means ultimately Prettier itself is responsible for calling `parse` in the appropriate plugins.

**Comment by thecrypticace** [comment]
@or2e Since you're using WSL2 — are your projects files under `/mnt/{drive-letter-here}` (e.g. `/mnt/c`)? WSL2 has serious performance issues accessing files shared with Windows via `/mnt/*` and any access of the filesystem slows things down significantly. I took your project, updated it to the latest version of the prettier plugin (v0.3.0) and ran the CLI against one of the files. It takes roughly 5.4s to process `./src/pages/clans/\[id\]/index.vue` under WebUI. About 4.7s of which is everything required to load your local copy of Tailwind, load and compile the config, and set up everything for processing. And the majority of this time is Node processing any `require()` calls.

If I move the repo so it's under `/home/wsl` the time drops significantly to 600ms total — with about 270ms of that being all the Tailwind, config loading, and setup. `require()` itself is slowed down a ton.

This might be something you can look at if you're still having performance problems.

@krisz094 Any chance you're using WSL as well?

**Comment by or2e** [comment]
@thecrypticace 
1/ The repository has been cloned into the distro (not mounted), `/root/WORK/crpg`
2/ 
`v0.2.8` - `~1500ms`
`v0.3.0` - `50-200ms`

ty!

**Comment by thecrypticace** [comment]
Fantastic!

**Comment by thecrypticace** [comment]
Hey all — since we've released v0.3.0, with the discovery regarding WSL and windows file system sharing, and that there's been little movement on this issue we're going to close it as the issues should be mostly solved.

If you're still experiencing problems and can provide a reproduction please feel free to open a new issue that we can take a look at.

Thanks everyone for their input! ✨

**Comment by ddahan** [comment]
Hi there. I'm curious to have extensive feedbacks from people who had the issue, an tested newer versions (0.3.0+) then. Thanks!
@gurgeous @kotaksempit @tjkohli @richardtallent-erm @Mattinton @Doesntmeananything @valgeirb @krisz094 @jd1378
